### PR TITLE
[enhancement] Rewrote World Utils

### PR DESCRIPTION
**Pull request by Avanatiker**
They are very useful yes.

### CODE REVIEW

**Comment by Avanatiker** [comment]
I hate codefactor :c always telling me iam too complex :P

File: src/main/kotlin/org/kamiblue/client/util/math/VectorUtils.kt [hunk 1]
```diff
@@ -63,6 +64,19 @@ object VectorUtils {
...
+        val yaw = -rotationRad.x - PI_FLOAT
+        val pitch = -rotationRad.y
+
+        val cosYaw = cos(yaw)
+        val sinYaw = sin(yaw)
+        val cosPitch = -cos(pitch)
+        val sinPitch = sin(pitch)
+
+        return Vec3d((sinYaw * cosPitch).toDouble(), sinPitch.toDouble(), (cosYaw * cosPitch).toDouble())
+    }
```
Comment (l1ving) [inline src/main/kotlin/org/kamiblue/client/util/math/VectorUtils.kt:78]: Could you leave a few comments explaining how this math works?

**Review by l1ving — CHANGES_REQUESTED** [review]
I'll have a couple more comments later but I'm a bit tired so I'll just leave the current ones for now

File: src/main/kotlin/org/kamiblue/client/module/modules/misc/AutoObsidian.kt [hunk 2]
```diff
@@ -66,7 +66,7 @@ internal object AutoObsidian : Module(
     private val searchShulker by setting("Search Shulker", false)
     private val leaveEmptyShulkers by setting("Leave Empty Shulkers", true, { searchShulker })
     private val autoRefill by setting("Auto Refill", false, { fillMode != FillMode.INFINITE })
-    private val instantMining by setting("Instant Mining", true)
+    private val instantMining by setting("Instant Mining", false)
```
[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/kotlin/org/kamiblue/client/module/modules/misc/AutoObsidian.kt:69]: Instant Mining doesn't work on 2b anymore 🗿

[the remark this answers is not in the document]
Comment (l1ving) [reply src/main/kotlin/org/kamiblue/client/module/modules/misc/AutoObsidian.kt:69]: It doesn't make sense being changed in this pull, they could at least explain why in the pull description.

[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/kotlin/org/kamiblue/client/module/modules/misc/AutoObsidian.kt:69]: > It doesn't make sense being changed in this pull, they could at least explain why in the pull description.

The utils changes are literally all about 2b so I don't see any problems

File: src/main/kotlin/org/kamiblue/client/util/math/VectorUtils.kt (hunk 1 above)
Comment (Xiaro) [reply src/main/kotlin/org/kamiblue/client/util/math/VectorUtils.kt:78]: It is just basic trigonometry math tho

File: src/main/kotlin/org/kamiblue/client/util/WorldUtils.kt [hunk 3]
```diff
@@ -84,7 +87,9 @@ object WorldUtils {
         for (x in xArray) {
             for (z in zArray) {
                 val result = rayTraceToGround(Vec3d(x, boundingBox.minY, z), stopOnLiquid)
-                results.add(result)
+                if (result != null) {
+                    results.add(result)
+                }
```
[the remark this answers is not in the document]
Comment (Xiaro) [reply src/main/kotlin/org/kamiblue/client/util/WorldUtils.kt:92]: let will make more lambda memes but ok

File: src/main/kotlin/org/kamiblue/client/util/math/VectorUtils.kt (hunk 1 above)
Comment (l1ving) [reply src/main/kotlin/org/kamiblue/client/util/math/VectorUtils.kt:78]: i'm tired

File: src/main/kotlin/org/kamiblue/client/module/modules/misc/AutoObsidian.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Avanatiker) [reply src/main/kotlin/org/kamiblue/client/module/modules/misc/AutoObsidian.kt:69]: It just should not be on one default. And i let it inside the code bc xiaro is sentimental :P

**Review by l1ving — APPROVED** [review]

File: src/main/kotlin/org/kamiblue/client/util/world/Interact.kt [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    val eyePos = player.getPositionEyes(1.0f)
+
+    return getVisibleSides(pos)
+        .minByOrNull { eyePos.squareDistanceTo(getHitVec(pos, it)) }
+}
+
+/**
+ * Get the "visible" sides related to player's eye position
+ *
+ * Reverse engineered from HauseMaster's anti cheat plugin
```
Comment (l1ving) [inline src/main/kotlin/org/kamiblue/client/util/world/Interact.kt:147]: Remove this line

**Review by l1ving — APPROVED** [review]

File: src/main/kotlin/org/kamiblue/client/module/modules/combat/CrystalAura.kt [hunk 5]
```diff
@@ -317,9 +318,10 @@ internal object CrystalAura : Module(
...
     }
 
-    private fun SafeClientEvent.getPlacePacket(pos: BlockPos, hand: EnumHand) =
-        CPacketPlayerTryUseItemOnBlock(pos, getHitSide(pos), hand, 0.5f, placeOffset, 0.5f)
-
+    private fun SafeClientEvent.getPlacePacket(pos: BlockPos, hand: EnumHand): CPacketPlayerTryUseItemOnBlock {
+        val side = getClosestVisibleSide(pos) ?: EnumFacing.UP
+        return CPacketPlayerTryUseItemOnBlock(pos, side, hand, 0.5f, placeOffset, 0.5f)
+    }
     private fun SafeClientEvent.packetExplode(entityID: Int, pos: BlockPos, vec3d: Vec3d) {
```
Comment (l1ving) [inline src/main/kotlin/org/kamiblue/client/module/modules/combat/CrystalAura.kt:325]:
```suggestion
    }

    private fun SafeClientEvent.packetExplode(entityID: Int, pos: BlockPos, vec3d: Vec3d) {
```

File: src/main/kotlin/org/kamiblue/client/util/world/Interact.kt [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
```
Comment (l1ving) [inline src/main/kotlin/org/kamiblue/client/util/world/Interact.kt:145]:
```suggestion
```

**Review by l1ving — CHANGES_REQUESTED** [review]

**Review by l1ving — APPROVED** [review]

**Comment by Xiaro** [comment]
Already tested on 2b

**Pull request merged**